oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                            padding                            |
+|                         invocation_id                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         message_length                        |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             length            |             flags             |
+|                          frame_number                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          body_length                          |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |                                                               |
 +                         body bytes...                         +
 |                                                               |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
 ```
 
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"padding:32,length:16,flags:16,body bytes...:64" -->
+<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
+"invocation_id:32,message_length:32,frame_number:32,body_length:32,body bytes...:64" -->
 
-- `padding`, 4 bytes
+- `invocation_id`, u32, little endian
 
-  For future use.
+  The unique ID of the invocation that the frame's forms a part of. All the
+  frames that make up a message MUST have the same invocation ID. The invocation
+  ID of a response message MUST match the value of its related request message.
+  The invocation ID MUST be incremented for each method invocation, naturally
+  wrapping when the maximum value for the data type is reached.
 
-- `length`, u16, little endian
+- `message_length`, unsigned 32-bit little-endian integer
 
-  Length of the entire frame, including this field. The length of a frame MUST
-  NOT exceed 4,000 bytes.
+  The length of the overall message. This value MUST be larger or equal to
```

Comment:
Is it about the upper level message length or the frame length?

Reply:
The "message length" is the length of the message, not the frame.